### PR TITLE
 [fix] #3516: make FindTrigger queries return original WASM

### PR DESCRIPTION
**Pull request by QuentinI**
## Description

Retain original WASM blobs in `TriggerSet`
Substitute internal representation for original when querying for Triggers
Remove internal representation from data_model

### Linked issue

Closes https://github.com/hyperledger/iroha/issues/3516

### Benefits

No more leaking internal representation
FindTrigger queries return actually useful WASM

### Checklist

- [x] I've read `CONTRIBUTING.md`
- [x] I've used the standard signed-off commit format (or will squash just before merging)
- [x] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: core/src/smartcontracts/isi/triggers/set.rs [hunk 1]
```diff
@@ -39,7 +40,88 @@ pub enum Error {
...
+#[derive(Clone, Debug)]
+pub struct LoadedAction<F> {
+    executable: LoadedExecutable,
+    pub repeats: Repeats,
+    pub authority: AccountId,
+    pub filter: F,
+    pub metadata: Metadata,
+}
+
+pub trait LoadedActionTrait {
```
Comment (mversic) [inline core/src/smartcontracts/isi/triggers/set.rs:52]: I don't think there is any need to have this trait

File: data_model/src/lib.rs [hunk 2]
```diff
@@ -1469,42 +1429,23 @@ impl TryFrom<f64> for Value {
...
+//     type Error = ErrorTryFromEnum<Value, Self>;
+//
+//     fn try_from(value: Value) -> Result<Self, Self::Error> {
+//         if let Value::Identifiable(IdentifiableBox::Trigger(trigger)) = value {
+//             return Ok(trigger);
+//         }
+//
+//         Err(Self::Error::default())
+//     }
+// }
```
Comment (mversic) [inline data_model/src/lib.rs:1448]: instead of commenting out everything, I would recommend you to use `unimplemented` in the body of the impl block. That way you will not accidentally forget to uncomment

File: data_model/src/trigger.rs [hunk 3]
```diff
@@ -317,11 +293,8 @@ pub mod action {
     /// Trait for common methods for all [`Action`]'s
     #[cfg(feature = "transparent_api")]
     pub trait ActionTrait {
```
Comment (mversic) [inline data_model/src/trigger.rs:295]: I don't think this trait is needed. But if it is, please move it's definition into `iroha_core`

**Review by mversic — CHANGES_REQUESTED** [review]
looks good overall

File: core/src/smartcontracts/isi/triggers/set.rs (hunk 1 above)
Comment (QuentinI) [reply core/src/smartcontracts/isi/triggers/set.rs:52]: It's used because `triggers::Set` stores `Action`s in [specialized maps](https://github.com/hyperledger/iroha/blob/107545cf45af1d457649f1dac63b4f6f2fac02a7/core/src/smartcontracts/isi/triggers/set.rs#L134) but allows to inspect [them regardless of filter type](https://github.com/hyperledger/iroha/blob/107545cf45af1d457649f1dac63b4f6f2fac02a7/core/src/smartcontracts/isi/triggers/set.rs#L554)
I don't really understand why `Action` needs to be parametrized by filter type at all and not have an enum instead, but unless we change that, I don't think there's an easy way to get rid of this trait.

File: data_model/src/lib.rs (hunk 2 above)
Comment (QuentinI) [reply data_model/src/lib.rs:1448]: I meant to delete these altogether, they aren't needed anymore

File: core/src/smartcontracts/isi/triggers/set.rs (hunk 1 above)
Comment (mversic) [reply core/src/smartcontracts/isi/triggers/set.rs:52]: after you have undrafted the PR, if you can remove this trait and replace trait objects with concrete types fine. If not, also fine. It's not a big deal

Comment (QuentinI) [reply core/src/smartcontracts/isi/triggers/set.rs:52]: Opened https://github.com/hyperledger/iroha/issues/3986

File: data_model/src/query/mod.rs [hunk 4]
```diff
@@ -1025,15 +1025,15 @@ pub mod trigger {
     }
 
     impl Query for FindTriggerById {
```
Comment (mversic) [inline data_model/src/query/mod.rs:1027]:
```suggestion
    impl Query for FindTriggerByHash {
```

and also change it's definition to:
```rs
        pub struct FindTriggerByHash {
            /// The Identification of the trigger to be found.
            pub id: EvaluatesTo<HashOf<Trigger<FilterBox>>>,
        }
```

Comment (mversic) [reply data_model/src/query/mod.rs:1027]: I believe we can also change `FindAllActiveTriggerIds` into `FindAllActiveTriggerHashes`

File: data_model/src/query/mod.rs [hunk 5]
```diff
@@ -1025,15 +1025,15 @@ pub mod trigger {
...
-        type Output = Trigger<TriggeringFilterBox, OptimizedExecutable>;
+        type Output = Trigger<TriggeringFilterBox>;
     }
 
     impl Query for FindTriggerKeyValueByIdAndKey {
```
Comment (mversic) [inline data_model/src/query/mod.rs:1031]: why do we have this query?

Comment (QuentinI) [reply data_model/src/query/mod.rs:1031]: I tracked it down to https://github.com/hyperledger/iroha/issues/1986, couldn't find any further motivation
I'd imagine it's what you would use if you receive `ExecuteTriggerEvent` and want to inspect the trigger for example

File: configs/peer/executor.wasm
Comment (mversic) [inline configs/peer/executor.wasm:1]: wow, what happened here? the size doubled somehow

**Review by mversic — DISMISSED** [review]

File: configs/peer/executor.wasm
Comment (QuentinI) [reply configs/peer/executor.wasm:1]: It's a product of my environment probably, it's a bit specific.
I can't rebuild it on a usual machine, unfortunately.

Comment (mversic) [reply configs/peer/executor.wasm:1]: hm, but wasm shouldn't be platform specific?

Comment (Arjentix) [reply configs/peer/executor.wasm:1]: One of us can build it on our system and push to your branch

Comment (mversic) [reply configs/peer/executor.wasm:1]: > It's a product of my environment probably, it's a bit specific.

it's true, I tested it now. That's surprising

Comment (Arjentix) [reply configs/peer/executor.wasm:1]: Are you sure you have run `iroha_wasm_builder_cli` with `--optimize` option?

File: data_model/src/trigger.rs [hunk 6]
```diff
@@ -64,62 +64,38 @@ pub mod model {
     )]
     #[display(fmt = "@@{id}")]
     #[ffi_type]
-    pub struct Trigger<F, E> {
+    pub struct Trigger<F> {
         /// [`Id`] of the [`Trigger`].
         pub id: TriggerId,
```
Comment (mversic) [inline data_model/src/trigger.rs:69]: can you go to `TriggerId` definition and swap `name` and `domain_id`. So that `Ord` impl orders first by domain then by name

File: flake.nix
Comment (Arjentix) [inline flake.nix:1]: Why did you change this file? It should be in a separate meaningful commit, I believe

File: configs/peer/executor.wasm
Comment (QuentinI) [reply configs/peer/executor.wasm:1]: > hm, but wasm shouldn't be platform specific?

Linker problem probably

File: core/src/smartcontracts/isi/triggers/set.rs [hunk 7]
```diff
@@ -507,32 +633,52 @@ impl Set {
...
+        };
+
+        let blob_hash = match event_type {
+            TriggeringEventType::Data => self
+                .data_triggers
+                .remove(id)
+                .map(|trigger| match trigger.executable {
+                    LoadedExecutable::Wasm(LoadedWasm { blob_hash, .. }) => Some(blob_hash),
+                    _ => None,
+                })
```
Comment (Arjentix) [inline core/src/smartcontracts/isi/triggers/set.rs:647]: I think you can move this part to a separate function to not to repeat this code for every trigger type. Will be something like:

```rust
.map(extract_blob_hash)
```

File: configs/peer/executor.wasm
Comment (QuentinI) [reply configs/peer/executor.wasm:1]: > Are you sure you have run `iroha_wasm_builder_cli` with `--optimize` option?

That is indeed the problem, thank you
I had to bypass parts of `iroha_wasm_builder` and inadvertently disabled optimization pass.

File: flake.nix
Comment (QuentinI) [reply flake.nix:1]: Accidental autoformat, will revert

**Review by mversic — APPROVED** [review]

File: configs/peer/executor.wasm
Comment (QuentinI) [reply configs/peer/executor.wasm:1]: Rebuilt with optimization pass, size back to normal.

**Review by mversic — APPROVED** [review]

**Review by mversic — APPROVED** [review]

**Review by Arjentix — APPROVED** [review]
